### PR TITLE
add multi-label training

### LINKED ISSUE #5560: Multi-label training
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

Hi, 
I really appreciate your great work for the cv community. Recently I realized the multi-label training and found that it worked well on my customized dataset. Specifically, I squeezed the labels before building the targets. The implementations are listed as follows. Any discussion is welcome.

    from collections import OrderedDict

    def squeeze_targets(self, targets):
        # input targets(image,class,x,y,w,h)

        # copy the input since we need to modify it
        targets_copy = targets.clone()
        # convert the key (image, x, y, w, h) to string for hashing
        keys = [np.array2string(key) for key in targets[:, [0, 2, 3, 4]].cpu().numpy()]

        # store all the classes for each unique box
        classes = OrderedDict()
        classes[keys[0]] = [targets[0, 1].long()]

        new_targets = [targets_copy[0]]
        for i, t in enumerate(targets[1:]):
            k = keys[i + 1]
            if k in classes.keys():
                classes[k].append(t[1].long())
            else:
                classes[k] = [t[1].long()]
                # here we change the class id to the row index of class array
                targets_copy[i + 1, 1] = len(new_targets)
                new_targets.append(targets_copy[i + 1])

        new_targets = torch.stack(new_targets)
        return new_targets, list(classes.values())

    def __call__(self, p, targets):  # predictions, targets, model
        device = targets.device
        lcls, lbox, lobj = torch.zeros(1, device=device), torch.zeros(1, device=device), torch.zeros(1, device=device)
        new_targets, classes = self.squeeze_targets(targets)
        tcls, tbox, indices, anchors = self.build_targets(p, new_targets)  # targets

        # Losses
        for i, pi in enumerate(p):  # layer index, layer predictions
            b, a, gj, gi = indices[i]  # image, anchor, gridy, gridx
            tobj = torch.zeros_like(pi[..., 0], device=device)  # target obj

            n = b.shape[0]  # number of targets
            if n:
                ps = pi[b, a, gj, gi]  # prediction subset corresponding to targets

                # Regression
                pxy = ps[:, :2].sigmoid() * 2. - 0.5
                pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]
                pbox = torch.cat((pxy, pwh), 1)  # predicted box
                iou = bbox_iou(pbox.T, tbox[i], x1y1x2y2=False, CIoU=True)  # iou(prediction, target)
                lbox += (1.0 - iou).mean()  # iou loss

                # Objectness
                tobj[b, a, gj, gi] = (1.0 - self.gr) + self.gr * iou.detach().clamp(0).type(tobj.dtype)  # iou ratio

                # Classification
                if self.nc > 1:  # cls loss (only if multiple classes)
                    t = torch.full_like(ps[:, 5:], self.cn, device=device)  # targets
                    for j in range(n):
                        for c in classes[tcls[i][j]]:
                            t[j, c] = self.cp
                    lcls += self.BCEcls(ps[:, 5:], t)  # BCE

### Use case

The training labels go to something like that.
cls_1, x1, y1, w1, h1
cls_2, x1, y1, w1, h1
cls_3, x2, y2, w2, h2
cls_4, x2, y2, w2, h2

After squeezing, the labels go to
row_num1, x1, y1, w1, h1
row_num2, x2, y2, w2, h2

[row_num1] cls_1, cls_2
[row_num2] cls_3, cls_4

### Additional

Before modification, the MAP goes to something like 95% but the prediction scores are quite low because of labels conflicts. And then it goes to 97% and the prediction scores go back normal since we allow the targets to have multiple self.cp in the same row.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by zachluo**
It needs to add the multi-label flag when using detect.py

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced YOLOv5 with multi-label classification support.

### 📊 Key Changes
- Introduced `multi_label` boolean parameter throughout the codebase to enable multi-label classification support.
- Modified the non_max_suppression function in `detect.py` to accommodate multi-labeling.
- Added multi_label support in the `ComputeLoss` class within `utils/loss.py` by implementing a new method `squeeze_targets` to process multi-label targets during training.

### 🎯 Purpose & Impact
- **Purpose**: To enable the YOLOv5 model to support scenarios where multiple labels may be applicable to a single detection.
- **Impact**: Users can now train YOLOv5 on datasets with multi-label annotations, increasing the model's flexibility and applicability to more complex tasks. This is particularly useful in fields like medical image analysis or situations where objects can logically belong to more than one category.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@zachluo thanks for the PR! 

Do you know the speed/CUDA memory impacts of your updates? For example total training time and CUDA usage before and after this PR on a common dataset like VOC, COCO, etc?

**Comment by zachluo** [comment]
> @zachluo thanks for the PR!
> 
> Do you know the speed/CUDA memory impacts of your updates? For example total training time and CUDA usage before and after this PR on a common dataset like VOC, COCO, etc?

Not yet, but I will run them on my own machine and let you know if ready.

**Comment by zachluo** [comment]
hi @glenn-jocher 
I have already tested the speed and memory and there are two things we need to discuss:
1) CUDA memory and CPU memory do not change but the training speed will go from 3.0it/s to 1.2s/it due to many `for` loops I used. I tried to remove all the `for` loop by advanced python APIs but it still has to keep the for loop here.[https://github.com/zachluo/yolov5/blob/multi_label_training/utils/loss.py#L192](url) Now the speed is 2.7it/s and still below the original speed. Do you have a better solution?

2) I add a multi-label flag in train.py to keep the training speed on coco unchanged. Besides, I found that COCO also has a few multi-label annotations, and thus it has to use multi-label to ensure everything unchanged before merging. Am I correct?

**Comment by glenn-jocher** [comment]
@zachluo that's good on the CPU and CUDA side, but we definitely want to address the speed differences, as the current PR appears to introduce training bottlenecks. This isn't too surprising though, as every part of target assignment and loss calculation is heavily optimized for the absolute fastest speeds.

I think yes as you said refactor away any for loops with matrix operations if possible, and also we need a global flag to simply skip this logic entirely unless multi-label datasets are detected. The current multi-label strategy is just to repeat a box (or segment) with a new class, i.e.:

So we can scan a dataset in the dataloader init and classify it as multi_label or single_label by the presence of non-unique boxes.

```python
45 0.479492 0.688771 0.955609 0.5955
45 0.736516 0.247188 0.498875 0.476417
50 0.637063 0.732938 0.494125 0.510583  # first class
60 0.637063 0.732938 0.494125 0.510583  # second class (same label)
```

EDIT: I don't think COCO is natively multi-label. Where did you see that?

**Comment by zachluo** [comment]
@glenn-jocher 

Sure, I think COCO is a single-label dataset. But when I scan the whole training set, I found only one text file with duplicate boxes. So do we need to set a proportion like 10% to classify a dataset as multi-label or single-label?

coco/labels/train2017/000000472314.txt
56 0.500000 0.500000 1.000000 1.000000 (duplicate box)
73 0.065773 0.093021 0.129016 0.184667
77 0.467016 0.286917 0.256344 0.390292
15 0.491391 0.603365 0.738281 0.512521
57 0.500000 0.500000 1.000000 1.000000 (duplicate box)

I also found the `targets` have more duplicate boxes maybe because of augmentation. More than 10 cases in one epoch.

coco/images/train2017/000000093075.jpg
tensor([[49.00000,  0.00000,  0.50000,  0.22622,  1.00000,  0.45245], (duplicate box)
        [49.00000, 38.00000,  0.50000,  0.22622,  1.00000,  0.45245], (duplicate box)
        [49.00000,  0.00000,  0.70655,  0.85242,  0.33744,  0.29516],
        [49.00000, 31.00000,  0.59171,  0.93853,  0.32876,  0.12293]], device='cuda:0')

coco/images/train2017/000000269541.jpg
tensor([[9.00000e+00, 0.00000e+00, 1.55771e-01, 3.30644e-01, 3.11542e-01, 6.61289e-01],
        [9.00000e+00, 3.00000e+00, 1.55651e-01, 3.43911e-01, 3.11302e-01, 6.87821e-01],
        [9.00000e+00, 1.50000e+01, 7.66197e-01, 4.00603e-01, 4.67603e-01, 5.37676e-01],
        [9.00000e+00, 5.60000e+01, 6.80697e-01, 4.08352e-01, 6.38604e-01, 8.16705e-01], (duplicate box)
        [9.00000e+00, 5.70000e+01, 6.80697e-01, 4.08352e-01, 6.38604e-01, 8.16705e-01], (duplicate box)
        [9.00000e+00, 7.70000e+01, 9.17680e-01, 1.36674e-01, 1.64638e-01, 2.73347e-01],
        [9.00000e+00, 2.50000e+01, 2.49296e-01, 9.82869e-01, 2.24197e-01, 3.42592e-02]], device='cuda:0')

**Comment by glenn-jocher** [comment]
@zachluo oh wow, nice investigative work. Yes then we either want to add a manual --multi-label argument to train.py argparser or make an automated threshold to determine multi-labelness of a dataset, i.e. maybe if >1% of all boxes are duplicated label it multi-label.

**Comment by zachluo** [comment]
@glenn-jocher yes, but I reckon --multi-label argument will be more flexible and it has been done in the latest commit.

**Comment by glenn-jocher** [comment]
@zachluo right, I see the arguments in train and detect. This creates a risk of user error though as the user needs to set an argument twice at different points in time in two different places, which will not have a 100% efficiency. 

You can leave the arguments, but we want to handle as much as we can automatically to remove human error, so we should attach a multi-label attribute to the model here and then use this value automatically in detect.py:
https://github.com/ultralytics/yolov5/blob/0453b758e7ff645528ae52aa85228f3672ff7594/train.py#L246-L256

**Comment by zachluo** [comment]
@glenn-jocher yes, totally agree. I have updated the commit.

**Comment by zachluo** [comment]
@glenn-jocher hi, what could I do then?

**Comment by thuangb** [comment]
Hi @glenn-jocher and @zachluo, what is the current state of this PR? Is it usable now?

**Comment by glenn-jocher** [comment]
@zachluo @thuangb current state is PR is failing CI and has conflicts with master.

@zachluo please resolve the CI issues and merge master, thank you!

**Comment by abatiSon** [comment]
I am waiting for this issue. What is the latest situation?

**Comment by zachluo** [comment]
@glenn-jocher hi, I encountered this CI issue that the detect.py will load the predefined yolov5n.pt due to this command in the workflow file (python detect.py --weights ${{ matrix.model }}.pt --device $di). As you suggest, I attach a multi-label attribute to the model during training. Thus, the CI error you also mentioned was raised. 

I removed this command (python detect.py --weights ${{ matrix.model }}.pt --device $di) in my workflow file and all the CI testing succeed, but it is not the best way cause it will affect other PR if it is merged to the master branch. What should I do?

BTW, it requires a review before merging to the master branch.

**Comment by zachluo** [comment]
> I am waiting for this issue. What is the latest situation?

Still

**Comment by zachluo** [comment]
@glenn-jocher pls help check the merging. Thanks!

**Review by thuangb — CHANGES_REQUESTED** [review]
The code is good, however the code for nms still handle the multi_label case incorrectly. I suggest to change x[:, 5:6] to x[:, 5:] in line 702 and line 716 of file utils/general.py. Also, the file export.py does not has multi_label options too, someone should look into that

**Comment by zachluo** [comment]
> The code is good, however the code for nms still handle the multi_label case incorrectly. I suggest to change x[:, 5:6] to x[:, 5:] in line 702 and line 716 of file utils/general.py. Also, the file export.py does not has multi_label options too, someone should look into that

@thuangb Thanks for your suggestion! But x is not the format [center x, center y, width, height, obj_conf,  cls_conf] as you expect, x is now the format [x1, y1, x2, y2, cls_conf, cls] after #692-698 in general.py. So x[:, 5:6] is correct.

export.py does not need a multi_label option. What I have done in this PR does not affect the export since the origin master branch has supported multi_label during testing as well as exporting. I just optimized the training details for multi-labels.

**Comment by thuangb** [comment]
> > The code is good, however the code for nms still handle the multi_label case incorrectly. I suggest to change x[:, 5:6] to x[:, 5:] in line 702 and line 716 of file utils/general.py. Also, the file export.py does not has multi_label options too, someone should look into that
> 
> @thuangb Thanks for your suggestion! But x is not the format [center x, center y, width, height, obj_conf, cls_conf] as you expect, x is now the format [x1, y1, x2, y2, cls_conf, cls] after #692-698 in general.py. So x[:, 5:6] is correct.
> 
> export.py does not need a multi_label option. What I have done in this PR does not affect the export since the origin master branch has supported multi_label during testing as well as exporting. I just optimized the training details for multi-labels.

May be I has some misunderstanding but can I ask that for multi_label, there is one output with multiple cls or mullltiple outputs with one cls?

For example, assume the ground truth has one object with two labels [0, 0, 0, 0, 0, 1, 1, 0, 0] then the output will be [0, 0, 0, 0, conf, 1, 2] or  [0, 0, 0, 0, conf, 1] and [0, 0, 0, 0, conf, 2]?

**Comment by zachluo** [comment]
@thuangb it is [0, 0, 0, 0, conf, 1] and [0, 0, 0, 0, conf, 2]. You can see that from line 694-695 in general.py.

**Comment by thuangb** [comment]
> @thuangb it is [0, 0, 0, 0, conf, 1] and [0, 0, 0, 0, conf, 2]. You can see that from line 694-695 in general.py.

Then I think we should re-work the nms code to merge multiples output into one output but with multiple cls because when we use multi_label, we want [0, 0, 0, 0, conf, 1, 2] more than [0, 0, 0, 0, conf, 1] and [0, 0, 0, 0, conf, 2]. May be add some code to if two bbox has large IoU and the cls is different then we merge them together? What do you think @zachluo, @glenn-jocher.

**Comment by zachluo** [comment]
> > @thuangb it is [0, 0, 0, 0, conf, 1] and [0, 0, 0, 0, conf, 2]. You can see that from line 694-695 in general.py.
> 
> Then I think we should re-work the nms code to merge multiples output into one output but with multiple cls because when we use multi_label, we want [0, 0, 0, 0, conf, 1, 2] more than [0, 0, 0, 0, conf, 1] and [0, 0, 0, 0, conf, 2]. May be add some code to if two bbox has large IoU and the cls is different then we merge them together? What do you think @zachluo, @glenn-jocher.

Yes, I did the merging in my customized codes, but that is not the main point in this PR. The multi-label has been implemented in the origin master branch. It should be included in your customized post-processing codes.

**Review by thuangb — DISMISSED** [review]

**Comment by thuangb** [comment]
Oh I see, so I think the PR is good to go @glenn-jocher

**Comment by zachluo** [comment]
I have solved the conflicts and pls merge master. Thanks. @glenn-jocher

**Comment by wxupjack** [comment]
If need someone to review this pending merging? @glenn-jocher

**Comment by glenn-jocher** [comment]
@wxupjack i appreciate your offer! We carefully review all pull requests and will merge them as appropriate. Thank you for contributing!
